sopel/modules/emailcheck.py

```diff
@@ -0,0 +1,282 @@
+                                        'List of channels to +R after malicious attempt to reg. '
+                                        'May be empty.')
+
+def setup(bot):
+    bot.config.define_section('emailcheck', EmailCheckSection)
+
+@dataclass
+class Email:
+    user: str
+    domain: str
+    def get_address(self):
+        return f'{self.user}@{self.domain}'
+    def __str__(self):
+        return self.get_address()
+    def __post_init__(self):
+        self.domain = self.domain.lower()
+
+@dataclass
+class DomainInfo:
+    score: float
+    flag_disposable: bool
+    flag_recent_abuse: bool
+
+def alert(bot, alert_msg: str, log_err: bool = False):
+    for channel in config.emailcheck.warn_chans:
+        bot.say(alert_msg, channel)
+    if log_err:
+        LOGGER.error(alert_msg)
+
+def add_badmail(bot, email):
+    #Right now we're BADMAILing whole domains. This might change.
+    bot.write("NICKSERV", "badmail", "add", f'*@{email.domain}')
+
+def fdrop(bot, nick: str):
+    bot.write("NICKSERV", "fdrop", nick.lower())
+
+def gline_ip(bot, ip: str, duration: str):
+    bot.write("GLINE", f'*@{ip}', duration, KILL_STR)
+
+def gline_username(bot, nick: str, duration: str):
+    if known_user := bot.users.get(Identifier(nick)):
+        username = known_user.user.lower() # Should already be lowercase
+        if IRCCLOUD_USER_REGEX.match(username):
+            bot.write("GLINE", f'{username}@*', duration, KILL_STR)
+            return
+        else:
+            alert(bot, f"User {nick} had unexpected non-IRCCloud username {username}", true)
+    else:
+        alert(bot, f"Couldn't find irccloud uid/sid for {nick} to G-line!", true)
+    kill_nick(bot, nick) # Something went wrong with G-line, so fall back to /kill
+
+def kill_nick(bot, nick: str):
+    bot.write("KILL", nick.lower(), KILL_STR)
+
+def gline_strategy(bot, nick):
+    if (known_user := bot.users.get(Identifier(nick))):
+        if hasattr(known_user, "ip"):
+            ip = known_user.ip
+            exemption = get_exemption(ip)
+            if exemption:
+                if "irccloud" in exemption.lower():
+                    # IRCCloud special case: ban uid/sid
+                    return ["gline_username", known_user.user]
+                else: # Fully exempt, so no g-line
+                    return None
+            else: # No exemption
+                return ["gline_ip", ip]
+    else: # Fail safely
+        return None
+
+def gline_or_kill(bot, nick: str, duration: str):
+    if strategy := gline_strategy(bot, nick):
+        if strategy[0] == "gline_ip":
+            gline_ip(bot, strategy[1], duration)
+        elif strategy[0] == "gline_username":
+            gline_username(bot, strategy[1], duration)
+        else:
+            alert(bot, f"Unknown strategy {strategy} for nick {nick}", true)
+            kill_nick(bot, nick) # safest option
+    else:
+        kill_nick(bot, nick) # duration ignored
+
+def protect_chans(bot):
```

Comment:
Is there a way to automatically deprotect the protected channels, or will they have to do this themselves?

Reply:
Manual for now

Reply:
In that case maybe drop a message mentioning that the channels are going into safe mode in oper chat, and a reminder that they will need to be manually `-R`'d.

Reply:
Good idea, done